sympy/ntheory/residue_ntheory.py

```
@@ -770,11 +770,18 @@ def nthroot_mod(a, n, p, all_roots=False):
     a, n, p = as_int(a), as_int(n), as_int(p)
     if n == 2:
         return sqrt_mod(a, p, all_roots)
+    # For prime moduli, handle zero residue early to include 0 as a root
+    # and avoid incorrect processing in the multiplicative group of units.
+    if not isprime(p):
+        raise NotImplementedError("Not implemented for composite p")
+
+    amod = a % p
+    if amod == 0:
+        return [0] if all_roots else 0
```

[major] The new early return skips the validation that  used to perform. Before this change, inputs with , , or  would raise (or return ) via that helper. Now, when , we return / even for invalid arguments. For example,  and  both silently return , whereas the base branch raises . We need to keep the zero short-circuit but only after the existing validation has run (e.g. call  first, or explicitly replicate those checks) so we don’t weaken the function’s invariants.}

---

Please ignore this truncated comment; see the follow-up on the same line for the full explanation.